File: pytet_v0.4/CTetris.py
from matrix import *
from tetris import *

class CTetris(Tetris):

    def deleteFullLines(self):
        
        array = self.oScreen.get_array()
        temp = self.iScreenDy-1
        
        while temp >= 0:
            if not(0 in array[temp]):
                i = temp
                while (i > 0): #0
                    array[i] = array[i-1]
                    i -= 1
                array[0] = [1]*self.iScreenDw + [0]*self.iScreenDx + [1]*self.iScreenDw

            else:
                 pass   

            temp -= 1

        return
        
    

    def accept(self, key):
        self.state = TetrisState.Running

        if key >= '0' and key <= '6':
            if self.justStarted == False:
                self.deleteFullLines()
            self.iScreen = Matrix(self.oScreen)
            self.idxBlockType = int(key)
            self.idxBlockDegree = 0
            self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
            self.top = 0
            self.left = Tetris.iScreenDw + self.iScreenDx//2 - self.currBlk.get_dx()//2
            self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
            self.tempBlk1 = self.tempBlk.binary() + self.currBlk.binary() #
            self.justStarted = False
            print()
            if self.tempBlk1.anyGreaterThan(1):
                self.state = TetrisState.Finished
            
            
            self.tempBlk = self.tempBlk + self.currBlk
            self.oScreen = Matrix(self.iScreen)
            self.oScreen.paste(self.tempBlk, self.top, self.left)
            return self.state #

        elif key == 'q':
            pass
        elif key == 'a': # move left
            self.left -= 1
        elif key == 'd': # move right
            self.left += 1
        elif key == 's': # move down
            self.top += 1
        elif key == 'w': # rotate the block clockwise
            self.idxBlockDegree = (self.idxBlockDegree + 1) % 4
            self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
        elif key == ' ': # drop the block
            while not self.tempBlk1.anyGreaterThan(1):
                    self.top += 1
                    self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
                    self.tempBlk1 = self.tempBlk.binary() + self.currBlk.binary() 
        else:
            print('Wrong key!!!')
            
        self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
        self.tempBlk3 = self.tempBlk.binary() + self.currBlk.binary()

        if self.tempBlk3.anyGreaterThan(1):   ## 벽 충돌시 undo 수행
            if key == 'a': # undo: move right
                self.left += 1
            elif key == 'd': # undo: move left
                self.left -= 1
            elif key == 's': # undo: move up
                self.top -= 1
                self.state = TetrisState.NewBlock
            elif key == 'w': # undo: rotate the block counter-clockwise
                self.idxBlockDegree = (self.idxBlockDegree - 1) % Tetris.nBlockDegrees
                self.currBlk = Tetris.setOfBlockObjects[self.idxBlockType][self.idxBlockDegree]
            elif key == ' ': # undo: move up
                self.top -= 1
                self.state = TetrisState.NewBlock
            
            

        
        self.tempBlk = self.iScreen.clip(self.top, self.left, self.top+self.currBlk.get_dy(), self.left+self.currBlk.get_dx())
        self.tempBlk = self.tempBlk + self.currBlk
        self.oScreen = Matrix(self.iScreen)
        self.oScreen.paste(self.tempBlk, self.top, self.left)

        return self.state
    
        
        

File: cpptet_v0.5/CTetris.cpp
#include "CTetris.h"



bool isvalue(int *a, int val,int size){ //1차 배열일경우 
    
    for (int i = 0; i < size; i++){
        if(a[i] == val){
            return true;
        }
            
    }

    return false;


}


Matrix*** CTetris::setOfCBlockObjects;

void CTetris::CdeleteTetris(){


    
    for(int i=0; i < Tetris::nBlockTypes; i++){
        for(int j=0; j <Tetris::nBlockDegrees;j++ ){
            delete setOfCBlockObjects[i][j];
        }
    }

    for(int i=0; i < Tetris::nBlockTypes; i++){
        delete [] setOfCBlockObjects[i];
    }
    delete[] setOfCBlockObjects;

    //delete tempCBlk;
    delete iCScreen;
    delete oCScreen;



}




void CTetris::deleteFullLines(){
    int **array1;
    array1 = this->oCScreen->get_array();
    int xx = this->oCScreen->get_dx();
    int yy = this->oCScreen->get_dy();
    int temp = this->iScreenDy-1;
        
    while(temp >= 0){
        if(!(isvalue(array1[temp],0,xx))){//not(0 in array[temp]):
            int i = temp;
            while(i > 0){
                array1[i] = array1[i-1];
                i -= 1;
            }
            //delete[] array1[0];
            array1[0] = new int[xx];

            
            for(int x=0; x < iScreenDw; x++){
                array1[0][x] = 1;
            }

            for(int x=0; x < this->iScreenDx; x++){
                array1[0][x+iScreenDw] = 0;
            }

            for(int x=0; x < iScreenDw; x++){
                array1[0][x+iScreenDw+this->iScreenDx] = 1;
            }

            continue;
        }


        else{
            temp -= 1;
        }
    }

    


}



void CTetris::init(int* setofblock[],int type, int degree){

    Tetris::init(setofblock,type,degree);

    setOfCBlockObjects = new Matrix**[nBlockTypes];
	for (int i = 0; i < nBlockTypes; i++)
		setOfCBlockObjects[i] = new Matrix* [nBlockDegrees];
	for (int i = 0; i < nBlockTypes; i++)
	{
		for (int j = 0; j < nBlockDegrees; j++)
		{
			setOfCBlockObjects[i][j] = new Matrix(setofblock[i*nBlockDegrees + j],sqrt(len(setofblock[i*nBlockDegrees])-1),sqrt(len(setofblock[i*nBlockDegrees])-1));
            setOfCBlockObjects[i][j]->mulc(i+1);
		}
	}


    



}

TetrisState CTetris::accept(char key){

    if (key >= '0' and key <= '6'){
        if (this->justStarted == false){
            deleteFullLines();
        }
        delete this->iCScreen;
        this->iCScreen = new Matrix(this->oCScreen);
    }

    this->oScreen = this->oCScreen->binary();
    delete this->oCScreen;

    
    this->state = this->Tetris::accept(key);

    this->currCBlk = CTetris::setOfCBlockObjects[idxBlockType][idxBlockDegree];
    this->tempCBlk = this->iCScreen->clip(this->top, this->left, this->top+this->currCBlk->get_dy(), this->left+this->currCBlk->get_dx());
    this->tempCBlk = this->tempCBlk->add(this->currCBlk);

    this->oCScreen = new Matrix(this->iCScreen);
    this->oCScreen->paste((this->tempCBlk), this->top, this->left);
    delete this->tempCBlk;

    return this->state;

}


